### PR TITLE
BPTL Dashboard: Fixed tracking number truncation issue for In Transit file

### PR DESCRIPTION
**Pull request by abhinavjonnada82**
This PR addresses following issue:
-> Provides user with an option to download either XLSX or CSV In Transit file
-> XLSX file fixes the tracking number truncation issue
-> SheetJS CDN library is loaded up on selecting _Create .csv file_ page instead of initial app load
-> Utilized a lightweight version of the SheetJS CDN library
-> Optimized as required & re-used code
-> Changed 'Whole Bl' to 'WHOLE BL' https://github.com/episphere/biospecimen/issues/550

https://github.com/episphere/biospecimen/issues/502

Checklist:
- [X] Code cleanup
- [X] Check for ES Lint warnings
- [X] Verify test cases
- [X] Check for GIT conflicts 
- [ ] Verified unit tests pass with current changes [N/A]
- [ ] Any dependencies or modules  [N/A]
- [X] Attach PoC
- [X] Add notes

Notes:
[2023-07-14-In-Transit-XLSX-data-export.xlsx](https://github.com/episphere/biospecimen/files/12052548/2023-07-14-In-Transit-XLSX-data-export.xlsx)
[2023-07-14-In-Transit-CSV-data-export.csv](https://github.com/episphere/biospecimen/files/12052551/2023-07-14-In-Transit-CSV-data-export.csv)
[2023-07-14-BSI-data-export.csv](https://github.com/episphere/biospecimen/files/12052554/2023-07-14-BSI-data-export.csv)

Test Cases Verified with PoC attached: 

1.) Under In Transit section when create file button is clicked, should trigger a modal with option of XLSX or CSV file to download
Screenshot 2023-07-14 at 1 38 20 PM

2.) Upon XLSX selection, should download XLSX file & all 10 digits of tracking number should be displayed
Screenshot 2023-06-30 at 3 21 39 PM

3.) Without refreshing the page, download In Transit CSV file
Screenshot 2023-06-30 at 3 37 02 PM

4.) Without refreshing the page, download Receipted CSV File
Screenshot 2023-06-30 at 3 25 20 PM

### CODE REVIEW

File: src/pages/receipts/csvFileReceipt.js [hunk 1]
```diff
@@ -49,22 +58,59 @@ const csvFileReceiptTemplate = async (username, auth, route) => {
...
+                            </div>`
+    confirmFileSelection();
   })
 }
 
+const confirmFileSelection = () => {
+  const radios = document.querySelectorAll('input[name="fileFormat"]');
+  radios.forEach(radio => {
+    radio.addEventListener('click', async (e) => {
+      const radioVal = radio.value;
```
Comment (Gbarra9) [inline src/pages/receipts/csvFileReceipt.js:102]: I like the ease of the user flow. The user clicks on either radio button and the file gets downloaded to their browser.

File: src/pages/receipts/csvFileReceipt.js [hunk 2]
```diff
@@ -314,21 +362,83 @@ const downloadCSVfile = (items, csv, title) => {
...
+*/ 
+
+const processInTransitXLSXData = (inTransitItems) => {
+  const header = ['Ship Date', 'Tracking Number', 'Shipped from Site', 'Shipped from Location', 'Shipped Date & Time', 'Expected Number of Samples', 'Temperature Monitor', 'Box Number', 'Specimen Bag ID Type', 'Full Specimen IDs', 'Material Type'];
+  const inTransitData = [header, ...inTransitItems.map(row => Object.values(row))];
+  loadSheetJScdn(inTransitData);
+};
+
+/**
+ * Loads SheetJS CDN upon .XLSX selection & then triggers SheetJS to process upon script on load
```
Comment (Gbarra9) [inline src/pages/receipts/csvFileReceipt.js:381]: I remember that this was a good approach in loading CDNs as needed.

**Review by Gbarra9 — APPROVED** [review]
No suggestions on my end. The comments I left are positive things I saw when looking at the code.

**Review by Gbarra9 — APPROVED** [review]

File: src/pages/receipts/csvFileReceipt.js [hunk 3]
```diff
@@ -131,19 +178,20 @@ const updateInTransitMapping = (shippedBoxes) => {
...
-            dataHolder['shipDateTime'] = i[fieldToConceptIdMapping.shippingShipDate] != undefined ? convertISODateTime(i[fieldToConceptIdMapping.shippingShipDate]) : ``
-            dataHolder['numSamples'] = specimenBagSize.length // to get number of samples
-            dataHolder['tempMonitor'] = i[fieldToConceptIdMapping.tempProbe] == fieldToConceptIdMapping.yes ? `Yes` : `No`
-            dataHolder['BoxId'] = i[fieldToConceptIdMapping.shippingBoxId] != undefined ? i[fieldToConceptIdMapping.shippingBoxId] : ``
-            dataHolder['specimenBagId'] = bagKeys[index]
-            dataHolder['fullSpecimenIds'] = fullSpecimenIds
-            dataHolder['materialType'] = materialTypeMapping(fullSpecimenIds)
-            holdProcessedResult.push(dataHolder)
+        let dataHolder = {
+          shipDate: i[fieldToConceptIdMapping.shippingShipDate]?.split("T")[0] || '',
```
Comment (we-ai) [inline src/pages/receipts/csvFileReceipt.js:182]: Use a better name than `i` (eg `box` or `shippedBox`).

File: src/pages/receipts/csvFileReceipt.js [hunk 4]
```diff
@@ -161,7 +209,7 @@ const updateInTransitMapping = (shippedBoxes) => {
 const materialTypeMapping = (specimenId) => {
   const tubeId = specimenId.split(' ')[1]
   const materialTypeObject = {'0001':'Serum', '0002':'Serum', '0011':'Serum', '0012':'Serum', '0021':'Serum', 
-                              '0003':  'Whole Bl', '0004': 'Whole Bl', '0005': 'Whole Bl', '0013': 'Whole Bl', '0014' : 'Whole Bl', '0024' : 'Whole Bl',
+                              '0003':  'WHOLE BL', '0004': 'WHOLE BL', '0005': 'WHOLE BL', '0013': 'WHOLE BL', '0014' : 'WHOLE BL', '0024' : 'WHOLE BL',
```
Comment (we-ai) [inline src/pages/receipts/csvFileReceipt.js:212]: Can be better looking.

File: src/pages/receipts/csvFileReceipt.js [hunk 5]
```diff
@@ -263,34 +311,36 @@ const getVialTypesMappings = (filteredResult) => {
...
-  delete i[fieldToConceptIdMapping.collectionLocation]
-  delete i['Connect_ID']
-  delete i[fieldToConceptIdMapping.collectionId]
-  delete i[fieldToConceptIdMapping.dateWithdrawn]
-  delete i[fieldToConceptIdMapping.clinicalDateTimeDrawn]
-  delete i[fieldToConceptIdMapping.dateReceived]
-  delete i[fieldToConceptIdMapping.collectionType]
-  delete i[fieldToConceptIdMapping.discardFlag]
-  delete i[fieldToConceptIdMapping.deviationNotFound]
+  i['Study ID'] = 'Connect Study';
```
Comment (we-ai) [inline src/pages/receipts/csvFileReceipt.js:314]: I suggest using a clearer name than `i`.

File: src/pages/receipts/csvFileReceipt.js [hunk 6]
```diff
@@ -314,21 +363,83 @@ const downloadCSVfile = (items, csv, title) => {
...
+  else {
+    link.setAttribute("href", `data:text/plain;charset=utf-8,${encodeURIComponent(blob)}`);
+    link.setAttribute("download", `${getCurrentDate()}-${title}.csv`);
+  }
+
+  document.body.appendChild(link);
+  link.click(); // Trigger download
+  document.body.removeChild(link);
+
+  // Display success message
```
Comment (we-ai) [inline src/pages/receipts/csvFileReceipt.js:435]: Do we need to display an error message when any error happens in above steps?

**Review by we-ai — APPROVED** [review]
Looks good.
Commented on styles (code readability).

File: src/pages/receipts/csvFileReceipt.js (hunk 4 above)
Comment (abhinavjonnada82) [reply src/pages/receipts/csvFileReceipt.js:212]: In JavaScript, user cannot directly map the same value to multiple different keys within a single object. Unless I create multiple references/objects to the same value

File: src/pages/receipts/csvFileReceipt.js (hunk 6 above)
Comment (abhinavjonnada82) [reply src/pages/receipts/csvFileReceipt.js:435]: Its creating a HTML link & downloading a file should be a success operation for the most part. If there's an error that would be inside the data processed

File: src/pages/receipts/csvFileReceipt.js (hunk 4 above)
Comment (we-ai) [reply src/pages/receipts/csvFileReceipt.js]: I meant each lines in the object are pushed to right, making code reading not as smooth.

**Pull request merged**